ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinConstraintsRule.java

```diff
@@ -117,7 +117,7 @@ public void onMatch(RelOptRuleCall call) {
     // These boolean values represent corresponding left, right input which is potential FK
     boolean leftInputPotentialFK = topRefs.intersects(leftBits);
     boolean rightInputPotentialFK = topRefs.intersects(rightBits);
-    if (leftInputPotentialFK && rightInputPotentialFK && joinType == JoinRelType.INNER) {
+    if (leftInputPotentialFK && rightInputPotentialFK && (joinType == JoinRelType.INNER || joinType == JoinRelType.SEMI)) {
```

Comment:
We'll need to go over this one.  I think the logic here matches the previous logic since as you mention, the restriction is on Join and not HiveJoin.  This means that anything that has joinType of INNER applies to the new SEMI type.  

So either the rule is supposed to handle semijoins fine or we have a bug.  Will talk with you offline about this.

Reply:
The old "SemiJoin" class extended the "Join" class, so this rule would be triggered by our HiveSemiJoin. So I think the logic here matches the existing logic. Let's take a look at this together though.